vim-patch:8.1.1564, 1570, 1623, 1712 : sign column takes up space

Shougo opened the pull request:
Problem:    Sign column takes up space.  (Adam Stankiewicz)
Solution:   Optionally put signs in the number column. (Yegappan Lakshmanan,
            closes vim/vim#4555, closes vim/vim#4515)
https://github.com/vim/vim/commit/394c5d8870b15150fc91a4c058dc571fd5eaa97e

~~I have tried the porting.  But it does not work as well...~~

Any advice and improvements are is appreciated!

Shougo committed:
vim-patch:8.1.1564: sign column takes up space

Problem:    Sign column takes up space.  (Adam Stankiewicz)
Solution:   Optionally put signs in the number column. (Yegappan Lakshmanan,
            closes vim/vim#4555, closes vim/vim#4515)
https://github.com/vim/vim/commit/394c5d8870b15150fc91a4c058dc571fd5eaa97e

Shougo committed:
vim-patch:8.1.1564: sign column takes up space

Problem:    Sign column takes up space.  (Adam Stankiewicz)
Solution:   Optionally put signs in the number column. (Yegappan Lakshmanan,
            closes vim/vim#4555, closes vim/vim#4515)
https://github.com/vim/vim/commit/394c5d8870b15150fc91a4c058dc571fd5eaa97e

Shougo commented:
Hm.  It has SEGV when `signcolumn=number` is enabled.

erw7 commented:
> Any advice and improvements are is appreciated!

@Shougo I think the following changes need to be made.

```C
diff --git a/src/nvim/screen.c b/src/nvim/screen.c
index bbfa16b33..17384b033 100644
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -2856,7 +2856,7 @@ win_line (
           if (count > 0) {
               get_sign_display_info(false, wp, lnum, row,
                       startrow, filler_lines, filler_todo, count,
-                      &c_extra, &c_final, extra, &p_extra, &n_extra,
+                      &c_extra, &c_final, extra, sizeof(extra), &p_extra, &n_extra,
                       &char_attr, &draw_state, &sign_idx);
           }
       }
@@ -2876,7 +2876,7 @@ win_line (
             int count = win_signcol_count(wp);
             get_sign_display_info(true, wp, lnum, row,
                 startrow, filler_lines, filler_todo, count,
-                &c_extra, &c_final, extra, &p_extra, &n_extra,
+                &c_extra, &c_final, extra, sizeof(extra), &p_extra, &n_extra,
                 &char_attr, &draw_state, &sign_idx);
           } else if (row == startrow + filler_lines) {
             // Draw the line number (empty space after wrapping). */
@@ -4479,6 +4479,7 @@ static void get_sign_display_info(
     int *c_extrap,
     int *c_finalp,
     char_u *extra,
+    size_t extra_size,
     char_u **pp_extra,
     int *n_extrap,
     int *char_attrp,
@@ -4508,7 +4509,7 @@ static void get_sign_display_info(
         *c_finalp = NUL;
 
         if (nrcol) {
-          sprintf((char *)extra, "%-*s ", number_width(wp),
+          snprintf((char *)extra, extra_size, "%-*s ", number_width(wp),
               *pp_extra);
           *pp_extra = extra;
           *n_extrap = (int)STRLEN(*pp_extra);
@@ -4523,12 +4524,12 @@ static void get_sign_display_info(
           *n_extrap = symbol_blen +
             (win_signcol_width(wp) - mb_string2cells(*pp_extra));
 
-          assert(sizeof(extra) > (size_t)symbol_blen);
-          memset(extra, ' ', sizeof(*extra));
+          assert(extra_size > (size_t)symbol_blen);
+          memset(extra, ' ', extra_size);
           memcpy(extra, *pp_extra, symbol_blen);
 
           *pp_extra = extra;
-          *pp_extra[*n_extrap] = NUL;
+          (*pp_extra)[*n_extrap] = NUL;
         }
       }
       *char_attrp = sign_get_attr(text_sign, SIGN_TEXT);
```

Shougo commented:
@erw7 Thank you!  SEGV is fixed now.
But the tests are failed.

```
From test_signs.vim:
Executed Test_sign() in   0.000297 seconds
Executed Test_sign_change_type() in   0.000182 seconds
Executed Test_sign_completion() in   0.000260 seconds
Executed Test_sign_cursor_position() in   0.000331 seconds
    Skipped
Executed Test_sign_delete_buffer() in   0.000183 seconds
Executed Test_sign_funcs() in   0.000293 seconds
Executed Test_sign_group() in   0.000305 seconds
Executed Test_sign_id_autogen() in   0.000271 seconds
Executed Test_sign_invalid_commands() in   0.000277 seconds
Executed Test_sign_jump_func() in   0.000157 seconds
Executed Test_sign_lnum_adjust() in   0.000266 seconds
Executed Test_sign_memfailures() in   0.000260 seconds
    Skipped
Executed Test_sign_numcol() in   0.000164 seconds
Executed Test_sign_priority() in   0.000291 seconds
Executed Test_sign_undefine_still_placed() in   0.000267 seconds
Executed Test_sign_unplace() in   0.000177 seconds
Executed 16 tests in   0.177111 seconds
4 FAILED:
Found errors in Test_sign():
Caught exception in Test_sign(): Vim(new):E325: ATTENTION @ function RunTheTest[44]..Test_sign, line 42
Found errors in Test_sign_group():
Caught exception in Test_sign_group(): Vim(split):E325: ATTENTION @ function RunTheTest[44]..Test_sign_group, line 207
Found errors in Test_sign_jump_func():
Caught exception in Test_sign_jump_func(): Vim(edit):E325: ATTENTION @ function RunTheTest[44]..Test_sign_jump_func, line 5
Found errors in Test_sign_numcol():
function RunTheTest[44]..Test_sign_numcol line 11: Expected '=> 01234' but got '=>01234 '
function RunTheTest[44]..Test_sign_numcol line 24: Expected '=>01234' but got '01234  '
SKIPPED Test_sign_cursor_position(): cannot make screendumps
SKIPPED Test_sign_memfailures(): Nvim does not support test_alloc_fail()
```

I need to investigate the reason...

Shougo commented:
I cannot fix the last errors...

```
Found errors in Test_sign_numcol():
function RunTheTest[44]..Test_sign_numcol line 11: Expected '=> 01234' but got '=>01234 '
function RunTheTest[44]..Test_sign_numcol line 24: Expected '=>01234' but got '01234  '
SKIPPED Test_sign_cursor_position(): cannot make screendumps
SKIPPED Test_sign_memfailures(): Nvim does not support test_alloc_fail()
```

erw7 commented on src/nvim/screen.c:
```diff
@@ -2903,24 +2869,32 @@ win_line (
+                &c_extra, &c_final, extra, sizeof(extra),
+                &p_extra, &n_extra,
+                &char_attr, &draw_state, &sign_idx);
+          } else if (row == startrow + filler_lines) {
```
> I cannot fix the last errors...

I think the problem is that you are making this place an `else if` statement. You must include up to line 2940 in the `else` statement.

Shougo replied:
Thanks.

Shougo commented:
```
Found errors in Test_sign_numcol():
function RunTheTest[44]..Test_sign_numcol line 24: Expected '=>01234' but got '01234  '
SKIPPED Test_sign_cursor_position(): cannot make screendumps
SKIPPED Test_sign_memfailures(): Nvim does not support test_alloc_fail()
```

The only one error exists...

Shougo commented:
Fixed lint errors.

erw7 commented on src/nvim/screen.c:
```diff
@@ -4494,6 +4474,83 @@ void screen_adjust_grid(ScreenGrid **grid, int *row_off, int *col_off)
+
+        if (nrcol) {
+          snprintf((char *)extra, extra_size, "%-*s ",
+                   number_width(wp), pp_extra);
```
You have a new enbug here when fixing the lint error.

```C 
snprintf((char *)extra, extra_size, "%-*s ",
              number_width(wp), *pp_extra);
```

Shougo replied:
Thanks.

erw7 commented:
> The only one error exists...

The following changes will make all the tests(`src/nvim/tesdir/test_signs.vim`, `test/functional/ui/sign_spec.lua`) pass.

```diff
diff --git a/src/nvim/option.c b/src/nvim/option.c
index 2564f1710..3fe7beefb 100644
--- a/src/nvim/option.c
+++ b/src/nvim/option.c
@@ -7397,7 +7397,8 @@ int win_signcol_count(win_T *wp)
   int maximum = 1, needed_signcols;
   const char *scl = (const char *)wp->w_p_scl;

-  if (*scl == 'n') {
+  if (*scl == 'n' && (*(scl + 1) == 'o'
+                      || (*(scl + 1) == 'u' && (wp->w_p_nu || wp->w_p_rnu)))) {
     return 0;
   }
   needed_signcols = buf_signcols(wp->w_buffer);
diff --git a/src/nvim/screen.c b/src/nvim/screen.c
index 70df2fd36..e4bf98957 100644
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -2855,7 +2855,7 @@ win_line (
           /* Show the sign column when there are any signs in this
            * buffer or when using Netbeans. */
           int count = win_signcol_count(wp);
-          if (count > 0 && !is_sign_number) {
+          if (count > 0) {
               get_sign_display_info(
                   false, wp, lnum, row,
                   startrow, filler_lines, filler_todo, count,
```

Shougo commented:
Thank you!  The tests are passed!

sentriz commented:
nice! thank you Shougo and erw7

erw7 committed:
vim-patch:8.1.1712: signs in number column cause text to be misaligned

Problem:    Signs in number column cause text to be misaligned.
Solution:   Improve alignment. (Yasuhiro Matsumoto, closes vim/vim#4694)
https://github.com/vim/vim/commit/d6bcff457799e491c3d27880858ec08e758e1849

erw7 committed:
vim-patch:8.1.1623: display wrong with signs in narrow number column

Problem:    Display wrong with signs in narrow number column.
Solution:   Increase the numbercolumn width if needed. (Yegappan Lakshmanan,
            closes vim/vim#4606)
https://github.com/vim/vim/commit/e4b407f536ba8bd007152649a347a95320d80fce

Shougo commented:
The failed reason is...

```
From test_textformat.vim:

Executed Test_format_align() in   0.001552 seconds

Executed Test_format_list_auto() in   0.001123 seconds

Executed Test_format_para() in  cat: write error: Resource temporarily unavailable

Makefile:181: recipe for target 'newtests' failed

make[1]: *** [newtests] Error 1

make[1]: Leaving directory '/home/travis/build/neovim/neovim/src/nvim/testdir'

Makefile:122: recipe for target 'oldtest' failed

make: *** [oldtest] Error 2
```

Too bad.

erw7 commented on src/nvim/option.c:
```diff
@@ -7397,7 +7404,9 @@ int win_signcol_count(win_T *wp)
   const char *scl = (const char *)wp->w_p_scl;
 
-  if (*scl == 'n') {
+  if (*scl == 'n'
```
You might want to add a comment.

erw7 commented on src/nvim/screen.c:
```diff
@@ -2816,6 +2816,7 @@ win_line (
   for (;; ) {
     int has_match_conc = 0;  ///< match wants to conceal
     bool did_decrement_ptr = false;
+
```
It is an unnecessary blank line.

erw7 commented on src/nvim/screen.c:
```diff
@@ -4494,6 +4473,88 @@ void screen_adjust_grid(ScreenGrid **grid, int *row_off, int *col_off)
+// If 'nrcol' is TRUE, the sign is going to be displayed in the number column.
+// Otherwise the sign is going to be displayed in the sign column.
+static void get_sign_display_info(
+    int nrcol,
```
I think `bool` is better.

erw7 commented on src/nvim/screen.c:
```diff
@@ -7508,3 +7577,5 @@ win_T *get_win_by_grid_handle(handle_T handle)
   }
   return NULL;
 }
+
```
It is an unnecessary blank line.

erw7 commented on src/nvim/testdir/test_signs.vim:
```diff
@@ -6,6 +6,8 @@ endif
 
 source screendump.vim
 
+set noswapfile
```
It's not present in the original vim patch, is this necessary? In my environment, the test passes without it.

Shougo replied:
I don't pass... Because

```
Found errors in Test_sign():
Caught exception in Test_sign(): Vim(new):E325: ATTENTION @ function RunTheTest[44]..Test_sign, line 42
Found errors in Test_sign_group():
Caught exception in Test_sign_group(): Vim(split):E325: ATTENTION @ function RunTheTest[44]..Test_sign_group, line 207
Found errors in Test_sign_jump_func():
Caught exception in Test_sign_jump_func(): Vim(edit):E325: ATTENTION @ function RunTheTest[44]..Test_sign_jump_func, line 5
```

I think swapfile is on in neovim default.

erw7 replied:
> I don't pass... 

I think that's because the swap file is still there when the test ended abnormally by `SEGV`. Therefore, instead of `setting noswapfile` here, you should delete any remaining swap files in your environment.

Shougo replied:
OK.  I have removed it.

erw7 commented on src/nvim/testdir/test_signs.vim:
```diff
@@ -1742,3 +1744,117 @@ func Test_sign_cursor_position()
+  " Enable number column. Check whether sign is displayed in the number column
+  set number
+  redraw!
+  call assert_equal("=>  01234", s:ScreenLine(1, 1, 9))
```
For some reason there seems to be one too many space.

https://ci.appveyor.com/project/neovim/neovim/builds/34075588/job/ao12eot4yxwssrip#L13543
```
From test_signs.vim:
Found errors in Test_sign_numcol():
function RunTheTest[44]..Test_sign_numcol line 30: Expected '=>  01234' but got ' => 01234'
../../../build/bin/nvim.exe -u NONE -U NONE -i NONE --noplugin --headless -S summarize.vim messages
Test results:
```

erw7 replied:
It seems that https://github.com/vim/vim/commit/4dff4aed09d2b0d570ca0d19de9cb08bdf03e695 is missing. I don't think this is necessary, but it seems to work differently with or without it compared to vim.

Shougo replied:
OK.  I have changed the test code.

Shougo committed:
vim-patch:8.1.1570: icon signs not displayed properly in the number column

Problem:    Icon signs not displayed properly in the number column.
Solution:   Display them properly. (Yegappan Lakshmanan, closes vim/vim#4559)
https://github.com/vim/vim/commit/4dff4aed09d2b0d570ca0d19de9cb08bdf03e695

Shougo commented:
Finally, all tests are passed!

Shougo commented:
I have resolved all comments and all tests are passed.
Can you merge it?

antoinemadec commented:
@erw7 , @marvim , is there anything blocking this merge?

teto commented:
@antoinemadec it would be nice to have your feedback too, did it work for you ? how long have you tested it ? Code reviews are nice but it's helpful to have user feedback too.

sentriz commented:
just did some testing. is working well for me :)

another thing though, `:echo has("patch-8.1.1564")` returns `0`. should it be 1?

fsouza commented:
@Shougo can you rebase this branch? If so, I can use for the next couple of days to confirm it works 😁 

If not, I can create a custom branch on my fork too, no worries.

jamessan commented:
> another thing though, :echo has("patch-8.1.1564") returns 0. should it be 1?

Not until after we finish merging the 8.0 patches.

antoinemadec commented:
> @antoinemadec it would be nice to have your feedback too, did it work for you ? how long have you tested it ? Code reviews are nice but it's helpful to have user feedback too.

@teto , I did today. Here are the main things I tried:

- setup: a lot of plugins (including coc.nvim).
- create git diffs and LSP errors on the same line: each creating signs
- use multiple tabs, cycle though them fast
- resizing my terminal

It seemed to work just fine for me :smiley: 

Cheers,
Antoine

Shougo commented:
> @Shougo can you rebase this branch? If so, I can use for the next couple of days to confirm it works grin

I have rebased the branch now.

fsouza commented:
@Shougo thanks! I created a branch earlier today and everything's been working fine here :D

fsouza commented:
Just confirming: been using it over the last 3 days with the builtin lsp client and it works like a charm! 😁

antoinemadec commented:
@jamessan @teto @Shougo @erw7 , are there other PRs blocking before we can merge this one?
The hype is strong with this one :smile:

Shougo commented:
Thank you for the approve.

teto commented:
vim implementation is a bit ugly IMO and the result could be improved too
![2020-07-25-000244_601x316_scrot](https://user-images.githubusercontent.com/886074/89734355-e65daf80-da5b-11ea-83f4-04c7290e9a57.png)
It's easier to merge though until we get a better solution.

Shougo commented:
Thank you for the merge!

Shougo commented:
> vim implementation is a bit ugly IMO and the result could be improved too

We need more active developers...
